Make KHARMA driver use MeshBlockPacks throughout

c-prather opened the pull request:
See #55.  If you get some time & the tests have passed, maybe try this @vedantdhruv96

Several of the `MeshDoThing` implementations are still loops, we could be faster for small blocks bundling them together.  Also, the second sync could likely be faster for the case that packs are small (usually done on CPUs).

However, those changes carry further risks.  If this both passes all tests & recovers some speed for the KHARMA driver, let's just merge this into stable.

c-prather committed:
Make KHARMA driver use MeshBlockPacks throughout, clearer region names

c-prather committed:
Fix Mesh-based KHARMA driver w/Dirichlet boundaries

c-prather commented:
The single failing test is the restarting test, and barely at all (1e-10 difference after 1M, vs 1e-11 previously).  Still might be a race in boundary exchange somewhere.

vedantdhruv96 commented:
I can confirm that this PR does indeed provide a speedup for the `kharma` driver - both single meshblock and multiple meshblocks. For a single meshblock, I now see on average 22M ZCPS as compared to ~17M ZCPS earlier. The multi-meshblock case on the other hand experienced a 10x speedup - 10M ZCPS (4 meshblocks for a 288x128x128 torus problem) as compared to 1M ZCPS earlier.

I vote to merge this into stable!

c-prather commented:
Wild.  There should be some more easy gains to merge to dev soon, then, too.  Let me loosen the restart test and we can merge.

Commit:
Loosen restart test

When the KHARMA driver is modified to operate on MeshData,
it loses some fidelity when restarting vs initializing.
Loosen the test to accommodate this, for now
